### PR TITLE
new aggregatesFrom and aggregatedVia fields

### PR DESCRIPTION
**Pull request by SrdjanStevanetic**

### CODE REVIEW

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Concept.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public Concept(Concept copy) {
     if (copy.getCloseMatch() != null) this.closeMatch = new ArrayList<>(copy.getCloseMatch());
     if (copy.getInScheme() != null) this.inScheme = new ArrayList<>(copy.getInScheme());
     if (copy.getNotation() != null) this.notation = new HashMap<>(copy.getNotation());
-    if (copy.exactMatch != null) this.exactMatch = (new ArrayList<>(copy.exactMatch));
+    if (copy.getSameReferenceLinks() != null) this.exactMatch = (new ArrayList<>(copy.getSameReferenceLinks()));
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Concept.java]: the SameReferenceLinks is an utility method, the data needs to be copied to the exact match as only this is stored into the database. Please revert this change

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Organization.java [hunk 2]
```diff
@@ -56,7 +58,9 @@
   FOAF_MBOX,
   HAS_ADDRESS,
   IDENTIFIER,
-  SAME_AS
+  SAME_AS,
+  AGGREGATES_FROM,
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Organization.java]: the new fields need to be added to the specs https://docs.google.com/spreadsheets/d/1b1UN27M2eCia0L54di0KQY7KcndTq8-wxzwM4wN-8DU/edit#gid=416760076 and the fields ordering needs to be consistent wit the ordeer in the specs

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Organization.java [hunk 3]
```diff
@@ -95,8 +101,10 @@ public Organization(Organization copy) {
       this.geographicLevel = new HashMap<>(copy.getGeographicLevel());
     this.country = copy.getCountry();
     if (copy.getAddress() != null) this.hasAddress = new Address(copy.getAddress());
-    if (copy.sameAs != null) this.sameAs = (new ArrayList<>(copy.sameAs));
-    if (copy.language != null) this.language = (new ArrayList<>(copy.language));
+    if (copy.getSameReferenceLinks() != null) this.sameAs = (new ArrayList<>(copy.getSameReferenceLinks()));
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Organization.java]: is the the sameAs field which is stored into the database, please revert this change

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Place.java [hunk 4]
```diff
@@ -67,7 +66,7 @@ public Place(Place copy) {
     this.latitude = copy.getLatitude();
     this.longitude = copy.getLongitude();
     this.altitude = copy.getAltitude();
-    if (copy.sameAs != null) this.sameAs = (new ArrayList<>(copy.sameAs));
+    if (copy.getSameReferenceLinks() != null) this.sameAs = (new ArrayList<>(copy.getSameReferenceLinks()));
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/Place.java]: please revert this change

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/TimeSpan.java [hunk 5]
```diff
@@ -61,7 +60,7 @@ public TimeSpan(TimeSpan copy) {
       this.isNextInSequence = new ArrayList<>(copy.getIsNextInSequence());
     this.begin = copy.getBeginString();
     this.end = copy.getEndString();
-    if (copy.sameAs != null) this.sameAs = (new ArrayList<>(copy.sameAs));
+    if (copy.getSameReferenceLinks() != null) this.sameAs = (new ArrayList<>(copy.getSameReferenceLinks()));
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/definitions/model/TimeSpan.java:62]: please revert this change

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/vocabulary/EntityFieldsTypes.java [hunk 6]
```diff
@@ -24,6 +24,8 @@ public enum EntityFieldsTypes {
   hiddenLabel(
       EntityFieldsTypes.FIELD_TYPE_KEYWORD, false, EntityFieldsTypes.FIELD_CARDINALITY_0_INFINITE),
   sameAs(EntityFieldsTypes.FIELD_TYPE_URI, false, EntityFieldsTypes.FIELD_CARDINALITY_0_INFINITE),
+  aggregatesFrom(EntityFieldsTypes.FIELD_TYPE_URI, false, EntityFieldsTypes.FIELD_CARDINALITY_0_INFINITE),
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/vocabulary/EntityFieldsTypes.java]: this field is not stored in the databse, must not be included here

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/web/xml/model/XmlOrganizationImpl.java [hunk 7]
```diff
@@ -63,13 +64,21 @@
       XML_HAS_ADDRESS,
       XML_IDENTIFIER,
       XML_SAME_AS,
+      XML_AGGREGATES_FROM,
```
Comment (gsergiu) [inline entity-management-definitions/src/main/java/eu/europeana/entitymanagement/web/xml/model/XmlOrganizationImpl.java]: order should be consistent with the specs, need to add the new fields to the specs

File: entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java [hunk 8]
```diff
@@ -166,6 +169,17 @@ public List<EntityRecord> findEntitiesByCoreference(
...
   }
   
+  public List<String> findAggregatesFrom(String entityId) {
+    Query<EntityRecord> query =
+        getDataStore()
+            .find(EntityRecord.class)
+            .disableValidation()
+            .filter(in(ENTITY_AGGREGATED_VIA, Arrays.asList(entityId)));
+    
+    List<EntityRecord> entityRecords = query.iterator(new FindOptions().projection().include(ENTITY_ID)).toList();
```
Comment (gsergiu) [inline entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java]: no iterator needed, the FindOptions should be provided to the find method

File: entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrOrganization.java [hunk 9]
```diff
@@ -1,23 +1,26 @@
...
+import eu.europeana.entitymanagement.vocabulary.EntitySolrFields;
+import eu.europeana.entitymanagement.vocabulary.OrganizationSolrFields;
 
 public class SolrOrganization extends SolrEntity<Organization> {
 
   @Field(EntitySolrFields.SAME_AS)
   private List<String> sameAs;
 
+  @Field(EntitySolrFields.AGGREGATED_VIA)
+  private List<String> aggregatedVia;
```
Comment (gsergiu) [inline entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrOrganization.java:22]: this filed is not in the specs for the search fields, need to clarify and eventually remove

File: entity-management-tests/src/integration-test/resources/solr-docker/conf/schema.xml [hunk 10]
```diff
@@ -197,6 +197,9 @@
 		<copyField source="owl_sameAs" dest="coref" />
 		<copyField source="skos_exactMatch" dest="coref" />
 
+		<field name="edm_aggregatedVia" type="string" indexed="false"
```
Comment (gsergiu) [inline entity-management-tests/src/integration-test/resources/solr-docker/conf/schema.xml]: need to clarify the specs firts, and eventually remove

File: entity-management-web/src/main/java/eu/europeana/entitymanagement/web/service/EntityRecordService.java [hunk 11]
```diff
@@ -141,6 +154,13 @@ public EntityRecord retrieveEntityRecord(String entityUri, boolean retrieveDisab
       throw new EntityRemovedException(
           String.format(EntityRecordUtils.ENTITY_ID_REMOVED_MSG, entityUri));
     }
+    
+    //for the organizations, populate the aggregatesFrom field
+    if(EntityTypes.Organization.getEntityType().equals(entityRecord.getEntity().getType())) {
```
Comment (gsergiu) [inline entity-management-web/src/main/java/eu/europeana/entitymanagement/web/service/EntityRecordService.java]: should avoid code dupplication, please create own method to fill the aggregate from

File: entity-management-definitions/src/main/java/eu/europeana/entitymanagement/vocabulary/EntityFieldsTypes.java (hunk 6 above)
Comment (SrdjanStevanetic) [reply entity-management-definitions/src/main/java/eu/europeana/entitymanagement/vocabulary/EntityFieldsTypes.java]: Every field that is defined in the entities, should be here, becuase in the fields validation, e.g. EntityFieldsDatatypeValidation, we call methods from the EntityFieldsTypes class, e.g. isMandatory(), which will throw an exception if the field is not defined. We validate all defined fields in the entities using reflection. See e.g. eu.europeana.entitymanagement.normalization.EntityFieldsDatatypeValidation.validateEntity

File: entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java (hunk 8 above)
Comment (SrdjanStevanetic) [reply entity-management-mongo/src/main/java/eu/europeana/entitymanagement/mongo/repository/EntityRecordRepository.java]: But the find method does not accept the FindOptions, right? All other methods use iterator as well.

File: entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrOrganization.java (hunk 9 above)
Comment (SrdjanStevanetic) [reply entity-management-solr/src/main/java/eu/europeana/entitymanagement/solr/model/SolrOrganization.java:22]: It is now.
